connectors/sources/slack.py

```diff
@@ -0,0 +1,429 @@
+                url = self._add_cursor(url, cursor)
+            response = await self._get_json(url)
+
+            for channel in response["channels"]:
+                if only_my_channels and not channel.get("is_member", False):
+                    continue
+                self._logger.debug(f"Yielding channel '{channel['name']}'")
+                yield channel
+            cursor = self._get_next_cursor(response)
+            if not cursor:
+                break
+
+    async def join_channel(self, channel_id):
+        url = f"{BASE_URL}/{ENDPOINTS[JOIN_CONVERSATIONS]}?channel={channel_id}"
+        return await self._get_json(url)
+
+    async def list_messages(self, channel_id, oldest, latest):
+        self._logger.info(
+            f"Fetching messages between {oldest} and {latest} from channel: '{channel_id}'"
+        )
+        message_batch_limit = 200
+        while True:
+            url = f"{BASE_URL}/{ENDPOINTS[CONVERSATION_HISTORY]}?channel={channel_id}&limit={message_batch_limit}&oldest={oldest}&latest={latest}"
+            response = await self._get_json(url)
+            for message in response.get("messages", []):
+                latest = message.get("ts")
+                if message.get("type") == "message":
+                    if message.get("reply_count", 0) > 0:
+                        async for thread_message in self.list_thread_messages(
+                            channel_id, message["ts"]
+                        ):
+                            yield thread_message
+                    else:
+                        yield message
+            if not response.get("has_more"):
+                break
+
+    async def list_thread_messages(self, channel_id, parent_ts):
+        message_batch_limit = 200
+        cursor = None
+        while True:
+            url = f"{BASE_URL}/{ENDPOINTS[CONVERSATION_REPLIES]}?channel={channel_id}&ts={parent_ts}&limit={message_batch_limit}"
+            if cursor:
+                url = self._add_cursor(url, cursor)
+            resonse = await self._get_json(url)
+            for message in resonse["messages"]:
+                if message["type"] == "message":
+                    yield message
+            cursor = self._get_next_cursor(resonse)
+            if not cursor:
+                break
+
+    async def list_users(self, cursor=None):
+        user_batch_limit = 200
+        while True:
+            url = f"{BASE_URL}/{ENDPOINTS[LIST_USERS]}?limit={user_batch_limit}"
+            if cursor:
+                url = self._add_cursor(url, cursor)
+            response = await self._get_json(url)
+            for member in response["members"]:
+                yield member
+            cursor = self._get_next_cursor(response)
+            if not cursor:
+                break
+
+    def _add_cursor(self, url, cursor):
+        return f"{url}&{CURSOR}={cursor}"
+
+    def _get_next_cursor(self, response):
+        return response.get(RESPONSE_METADATA, {}).get(NEXT_CURSOR)
```

Comment:
This can be inlined, IMO a little duplication does not hurt in this case

Reply:
I messed up the logic once and decided that I didn't want to have to retype this invocation multiple times and keep risking making another bug. What's the downside, in your opinion, to having this in a helper function?

Reply:
I think it's okay to have this function, I just find one-liners like this confusing me when reading/reviewing, thought train is: "hmmm it's a separate function, probably lots of logic inside". Also I personally don't like methods that only encapsulate field access. It's definitely okay to have this method like this, just being picky.

Reply:
ha, ok. If you don't feel strongly, I'm going to leave this one, so that the `, {}` doesn't get missed the next time we add page handling, or something.